### PR TITLE
Refactor Polymer 2.0 element scanner into phases

### PR DESCRIPTION
**Pull request by rictic**
- [x] CHANGELOG.md has been updated

This PR subsumes https://github.com/Polymer/polymer-analyzer/pull/609 

It should otherwise be a noop change, though I'd expect it to fix some weird bugs. It's a stepping stone to #563 

Conceptually the idea is that in one pass we identify all classes and potential custom elements, and in a separate visitor we collect up all `customElements.define` calls. We then produce ScannedPolymerElements from these two lists using pretty much the same logic as before.

### CODE REVIEW

File: src/polymer/polymer-element.ts [hunk 1]
```diff
@@ -168,6 +169,11 @@ export class ScannedPolymerElement extends ScannedElement implements
     if (options && options.methods) {
       options.methods.forEach((m) => this.addMethod(m));
     }
+    if (this.jsdoc) {
+      const summaryTag = jsdoc.getTag(this.jsdoc, 'summary');
```
Comment (FredKSchott) [inline src/polymer/polymer-element.ts]: `jsdoc.getTag(this.jsdoc, 'summary', 'description');` should give you the description you're trying to access below.

File: src/polymer/polymer2-element-scanner.ts [hunk 2]
```diff
@@ -436,66 +319,139 @@ class ElementVisitor implements Visitor {
...
+    this.classes.push({name, namespacedName, doc, ast});
+    if (ast.type === 'ClassExpression') {
+      this.alreadyMatched.add(ast);
+    }
+  }
+}
+
+interface ElementDefineCall {
+  tagName: TagNameExpression;
+  clazz: ElementClassExpression;
```
Comment (FredKSchott) [inline src/polymer/polymer2-element-scanner.ts]: putting the java in javascript :)

File: src/polymer/polymer2-element-scanner.ts [hunk 3]
```diff
@@ -65,161 +57,184 @@ export class Polymer2ElementScanner implements JavaScriptScanner {
...
-        className: namespacedClassName,
-        privacy: getOrInferPrivacy(namespacedClassName, docs, false)
+        description: (docs.description || '').trim(),
+        events: esutil.getEventComments(node),
+        sourceRange: document.sourceRangeForNode(node),
+        properties: getProperties(node, document),
+        methods: getMethods(node, document),
+        superClass: this._getExtends(node, docs, warnings, document),
+        mixins: jsdoc.getMixins(document, node, docs, warnings),
+        privacy: getOrInferPrivacy(className || '', docs, false),  //
```
Comment (FredKSchott) [inline src/polymer/polymer2-element-scanner.ts]: a wild `//`

File: src/polymer/polymer2-element-scanner.ts [hunk 4]
```diff
@@ -65,161 +57,184 @@ export class Polymer2ElementScanner implements JavaScriptScanner {
...
+          }
+          return astValue.getIdentifierName(m.key) === methodName;
+        });
+    if (observedAttributesDefn) {
+      const body = observedAttributesDefn.value.body.body[0];
+      if (body && body.type === 'ReturnStatement' && body.argument) {
+        return body.argument;
+      }
     }
+    return;
```
Comment (FredKSchott) [inline src/polymer/polymer2-element-scanner.ts]: unnecessary? (or is this a TS thing)

File: src/polymer/polymer-element.ts (hunk 1 above)
Comment (rictic) [inline src/polymer/polymer-element.ts]: Done

File: src/polymer/polymer2-element-scanner.ts (hunk 3 above)
Comment (rictic) [inline src/polymer/polymer2-element-scanner.ts]: Done

File: src/polymer/polymer2-element-scanner.ts (hunk 4 above)
Comment (rictic) [inline src/polymer/polymer2-element-scanner.ts]: Looks like unnecessary

File: src/polymer/polymer2-element-scanner.ts (hunk 2 above)
Comment (rictic) [inline src/polymer/polymer2-element-scanner.ts]: image

File: src/polymer/polymer2-element-scanner.ts [hunk 5]
```diff
@@ -65,161 +57,183 @@ export class Polymer2ElementScanner implements JavaScriptScanner {
...
-      // can detect registration by either name.
-      this._possibleElements.set(className, element);
-      this._possibleElements.set(namespacedClassName, element);
+  private _makeElementFeature(
+      element: CustomElementDefinition,
+      document: JavaScriptDocument): ScannedPolymerElement {
+    const node = element.clazz.ast;
+    const docs = element.clazz.doc;
+    const className = element.clazz.namespacedName;
+    let tagName;
```
Comment (FredKSchott) [inline src/polymer/polymer2-element-scanner.ts]: surprised ts isn't getting mad at this, my machine is registering it as an implicit any. Can you define a type?

File: src/polymer/polymer2-element-scanner.ts [hunk 6]
```diff
@@ -65,161 +57,183 @@ export class Polymer2ElementScanner implements JavaScriptScanner {
...
+    // TODO(rictic): support `@customElements explicit-tag-name` from jsdoc
+    if (element.definition &&
+        element.definition.tagName.type === 'string-literal') {
+      tagName = element.definition.tagName.value;
+    } else if (
+        node.type === 'ClassExpression' || node.type === 'ClassDeclaration') {
+      tagName = getIsValue(node);
```
Comment (FredKSchott) [inline src/polymer/polymer2-element-scanner.ts]: what if tagName is undefined? Is that allowed?

File: src/polymer/polymer2-element-scanner.ts [hunk 7]
```diff
@@ -436,66 +318,139 @@ class ElementVisitor implements Visitor {
...
+        esutil.getAttachedComment(parent) || '';
+    this._classFound(name, jsdoc.parseJsdoc(comment), node);
+  }
+
+  private _classFound(
+      name: string|undefined, doc: jsdoc.Annotation, ast: estree.Node) {
+    const namespacedName = name && getNamespacedIdentifier(name, doc);
+
+    this.classes.push({name, namespacedName, doc, ast});
+    if (ast.type === 'ClassExpression') {
```
Comment (FredKSchott) [inline src/polymer/polymer2-element-scanner.ts]: can you explain this, and why `enterClassDeclaration` doesn't check for alreadyMatched`, in documentation?

File: src/polymer/polymer2-element-scanner.ts [hunk 8]
```diff
@@ -436,66 +318,139 @@ class ElementVisitor implements Visitor {
...
+    const callee = astValue.getIdentifierName(node.callee);
+    if (!(callee === 'window.customElements.define' ||
+          callee === 'customElements.define')) {
+      return;
     }
-    return {successful: true, value: element.tagName};
+
+    const tagNameExpression =
+        node.arguments[0] && this._getTagNameExpression(node.arguments[0]);
+    if (tagNameExpression == null) {
```
Comment (FredKSchott) [inline src/polymer/polymer2-element-scanner.ts]: why `tagNameExpression == null` here, and `!elementClassExpression` below? Why not check falsey for both? Please document if there is a specific reason.

File: src/polymer/polymer2-element-scanner.ts [hunk 9]
```diff
@@ -436,66 +318,139 @@ class ElementVisitor implements Visitor {
...
+
+  enterVariableDeclarator(
+      node: estree.VariableDeclarator, parent: estree.Node) {
+    if (node.init) {
+      this.handleGeneralAssignment(
+          astValue.getIdentifierName(node.id), node.init, node, parent);
+    }
+  }
+
+  private handleGeneralAssignment(
```
Comment (FredKSchott) [inline src/polymer/polymer2-element-scanner.ts]: document, how does this handle a general assignment?

File: src/test/polymer/polymer2-element-scanner_test.ts [hunk 10]
```diff
@@ -371,6 +371,18 @@ namespaced name.`,
         mixins: ['Mixin'],
         warningUnderlines: [],
       },
+      {
```
Comment (FredKSchott) [inline src/test/polymer/polymer2-element-scanner_test.ts]: is the changed behavior still covered by current tests? There are a lot of edge cases handled above that should be tested as well.

File: src/polymer/polymer2-element-scanner.ts [hunk 11]
```diff
@@ -65,161 +57,183 @@ export class Polymer2ElementScanner implements JavaScriptScanner {
...
+        new CustomElementsDefineCallFinder(document);
+    // Find all classes and all calls to customElements.define()
+    await Promise.all([visit(classFinder), visit(elementDefinitionFinder)]);
+
+    const elementDefinitionsByClassName = new Map<string, ElementDefineCall>();
+    const elementDefinitionsByClassExpression =
+        new Map<estree.ClassExpression, ElementDefineCall>();
+
+    for (const defineCall of elementDefinitionFinder.calls) {
+      if (defineCall.clazz.type === 'MaybeChainedIdentifier') {
```
Comment (justinfagnani) [inline src/polymer/polymer2-element-scanner.ts]: can you add a comment for those who might be unfamiliar with MaybeChainedIdentifier

File: src/polymer/polymer2-element-scanner.ts [hunk 12]
```diff
@@ -65,161 +57,183 @@ export class Polymer2ElementScanner implements JavaScriptScanner {
...
-  }
-}
+    const classFinder = new ClassFinder();
+    const elementDefinitionFinder =
+        new CustomElementsDefineCallFinder(document);
+    // Find all classes and all calls to customElements.define()
+    await Promise.all([visit(classFinder), visit(elementDefinitionFinder)]);
+
+    const elementDefinitionsByClassName = new Map<string, ElementDefineCall>();
+    const elementDefinitionsByClassExpression =
```
Comment (justinfagnani) [inline src/polymer/polymer2-element-scanner.ts]: Why only `ClassExpression`s and not `ClassDeclaration`s?

File: src/polymer/polymer2-element-scanner.ts [hunk 13]
```diff
@@ -65,161 +57,183 @@ export class Polymer2ElementScanner implements JavaScriptScanner {
...
+        elementDefinitionsByClassExpression.set(defineCall.clazz, defineCall);
+      }
+    }
+    // TODO(rictic): emit ElementDefineCallFeatures for define calls that don't
+    //     map to any local classes?
+
+    const customElements: CustomElementDefinition[] = [];
+    const normalClasses = [];
+    for (const clazz of classFinder.classes) {
+      if (clazz.ast.type === 'ClassExpression') {
```
Comment (justinfagnani) [inline src/polymer/polymer2-element-scanner.ts]: There seems to be 3 separate cases here, it'd be nice to outline them with comments

File: src/polymer/polymer2-element-scanner.ts [hunk 14]
```diff
@@ -65,161 +57,183 @@ export class Polymer2ElementScanner implements JavaScriptScanner {
...
+    }
 
-class ElementVisitor implements Visitor {
-  private _possibleElements = new Map<string, ScannedElement>();
-  private _registeredButNotFound = new Map<string, TagNameExpression>();
-  private _elements: Set<ScannedElement> = new Set();
-  private _document: JavaScriptDocument;
-  // TODO(rictic): write a WarningFeature. Emit them from this scanner.
-  private _warnings: Warning[] = [];
+    const elementFeatures: ScannedPolymerElement[] = [];
```
Comment (justinfagnani) [inline src/polymer/polymer2-element-scanner.ts]: golf: `const elementFeatures = customElements.map((e) => this._makeElementFeature(e, document));`

File: src/polymer/polymer2-element-scanner.ts (hunk 6 above)
Comment (justinfagnani) [inline src/polymer/polymer2-element-scanner.ts]: Yes, for base classes

File: src/polymer/polymer2-element-scanner.ts (hunk 5 above)
Comment (justinfagnani) [inline src/polymer/polymer2-element-scanner.ts]: What version of tsc do you have?

File: src/polymer/polymer2-element-scanner.ts [hunk 15]
```diff
@@ -65,161 +57,183 @@ export class Polymer2ElementScanner implements JavaScriptScanner {
...
-      // can detect registration by either name.
-      this._possibleElements.set(namespacedClassName, element);
-      if (className) {
-        this._possibleElements.set(className, element);
+      if (observedAttributes != null) {
+        scannedElement.attributes = observedAttributes;
       }
+
+    } else {
+      scannedElement = new ScannedPolymerElement({
```
Comment (justinfagnani) [inline src/polymer/polymer2-element-scanner.ts]: can you comment on the cases that fall through here?

File: src/polymer/polymer2-element-scanner.ts [hunk 16]
```diff
@@ -65,161 +57,183 @@ export class Polymer2ElementScanner implements JavaScriptScanner {
...
+    return scannedElement;
   }
 
-  leaveVariableDeclaration(
-      node: estree.VariableDeclaration, _parent: estree.Node) {
-    if (this._currentElementNode === node) {
-      // Clean up state when we leave a declaration that defined an element.
-      this._currentElement = null;
-      this._currentElementNode = null;
+  private _getReturnValueOfStaticGetter(
```
Comment (justinfagnani) [inline src/polymer/polymer2-element-scanner.ts]: `getStaticGetterValue` already exists in polymer2-config.ts

File: src/polymer/polymer2-element-scanner.ts [hunk 17]
```diff
@@ -436,66 +318,139 @@ class ElementVisitor implements Visitor {
...
-      if (element) {
-        element.className = className;
-        const tagNameResult = this.getTagNameFromExpression(tagNameExpression);
-        if (tagNameResult.successful) {
-          element.tagName = tagNameResult.value;
-        } else {
-          this._warnings.push(tagNameResult.value);
-        }
-        this._elements.add(element);
+  ast: estree.Node;
```
Comment (justinfagnani) [inline src/polymer/polymer2-element-scanner.ts]: `ast` -> `node`

File: src/polymer/polymer2-element-scanner.ts [hunk 18]
```diff
@@ -436,66 +318,139 @@ class ElementVisitor implements Visitor {
...
+}
+class ClassFinder implements Visitor {
```
Comment (justinfagnani) [inline src/polymer/polymer2-element-scanner.ts]: add newline

File: src/polymer/polymer2-element-scanner.ts [hunk 19]
```diff
@@ -436,66 +318,139 @@ class ElementVisitor implements Visitor {
...
+
+interface ElementDefineCall {
+  tagName: TagNameExpression;
+  clazz: ElementClassExpression;
+}
+type ElementClassExpression = estree.ClassExpression|{
+  type: 'MaybeChainedIdentifier';
+  name: string, sourceRange: SourceRange
+};
+class CustomElementsDefineCallFinder implements Visitor {
```
Comment (justinfagnani) [inline src/polymer/polymer2-element-scanner.ts]: add some newlines

File: src/polymer/polymer2-element-scanner.ts [hunk 20]
```diff
@@ -436,66 +318,139 @@ class ElementVisitor implements Visitor {
...
+  readonly warnings: Warning[] = [];
+  readonly calls: ElementDefineCall[] = [];
+  private readonly _document: JavaScriptDocument;
+  constructor(document: JavaScriptDocument) {
```
Comment (justinfagnani) [inline src/polymer/polymer2-element-scanner.ts]: newlines!

File: src/polymer/polymer2-element-scanner.ts [hunk 21]
```diff
@@ -436,66 +318,139 @@ class ElementVisitor implements Visitor {
...
     }
-    return {successful: true, value: element.tagName};
+
+    const tagNameExpression =
+        node.arguments[0] && this._getTagNameExpression(node.arguments[0]);
+    if (tagNameExpression == null) {
+      return;
+    }
+    const elementClassExpression =
+        node.arguments[1] && this._getElementClassExpression(node.arguments[1]);
```
Comment (justinfagnani) [inline src/polymer/polymer2-element-scanner.ts]: might as well make `_getElementClassExpression` handle undefined

**Comment by justinfagnani** [comment]
Since you're adding some support for `@customElement` can you make sure you add it to the jsdoc parser config? Also comments on #605 that this does part of it.

**Comment by rictic** [comment]
Punting the jsdoc tag changes to a follow up
